### PR TITLE
Fix Android background service

### LINKED ISSUE #3999: App not syncing after 5 minutes of inactivity
## 🐛 Bug Report
App doesn't sync automatically when resuming after 5 minutes of inactivity.

## Context

Build/Branch: Giskard 1.7.1
Phone OS: Android
Phone model: any

## Expected behavior
App should sync automatically like Daneel.

## To reproduce
1. Be logged in, with app synced and go to background
2. Make some changes in the webapp, e.g: start a time entry
3. Wait 5 minutes and resume the app
4. 🐛 Sync should start and the changes should appear

## Comments
I double checked this only happens in Android, iOS works as expected.

### PR DESCRIPTION
**Pull request by juanlaube**
## What's this?
This fixes (or adds?) the detection of the app going to background and becoming foreground again. 

### Relationships

Closes #3999 

## Why do we want this?
Different parts of the app depend on this, for example, sync should happen when returning after 5 minutes in background.

## How is it done?
The new `ApplicationLifecycleObserver` is registered as application's activity lifecycle callbacks and component callbacks.
When the app goes to background, [memory is trimmed with `UiHidden` level](https://developer.android.com/reference/android/content/ComponentCallbacks2.html#TRIM_MEMORY_UI_HIDDEN), the `isInBackground` flag is set and we call `EnterBackground` in the background service.
When the app becomes foreground again, an activity is resumed, and if `isInBackground` is set, we call `EnterForeground` in background service.

### Why not another way?
Another approach is using Android Architecture Components, which includes a new `ProcessLifecycleOwner` class, but this means installing also a lot of new dependencies.

### Side effects
None.

## Review hints
Please test in real devices, put the app in background and return after 5 minutes, see if sync triggers.

## :squid: Permissions
Only after @heytherewill, @simonrozsival and @daividssilverio are ok with this.

### CODE REVIEW

File: Toggl.Giskard/Startup/ApplicationLifecycleObserver.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        {
+        }
+
+        public void OnActivityStopped(Activity activity)
+        {
+        }
+
+        public void OnConfigurationChanged(Configuration newConfig)
+        {
+            throw new NotImplementedException();
```
Comment (daividssilverio) [inline Toggl.Giskard/Startup/ApplicationLifecycleObserver.cs]: Throwing doesn't sound like a good idea here... could you just leave it empty?

Comment (juanlaube) [reply Toggl.Giskard/Startup/ApplicationLifecycleObserver.cs]: oh I forgot to remove this 🙈

Comment (dshr) [inline Toggl.Giskard/Startup/ApplicationLifecycleObserver.cs]: Do we really need this `throw`?

Comment (juanlaube) [reply Toggl.Giskard/Startup/ApplicationLifecycleObserver.cs]: No, see previous comment ☝️ 🤦‍♂️

Comment (dshr) [reply Toggl.Giskard/Startup/ApplicationLifecycleObserver.cs]: Nvm, missed the changes

File: Toggl.Giskard/Startup/Setup.cs [hunk 2]
```diff
@@ -136,6 +138,11 @@ protected override void InitializeApp(IMvxPluginManager pluginManager, IMvxAppli
 
             ensureDataSourceInitializationIfLoggedIn();
 
+            var mvxApplication = MvxAndroidApplication.Instance;
```
Comment (simonrozsival) [inline Toggl.Giskard/Startup/Setup.cs]: This block of code would deserve a separate method.

**Comment by daividssilverio** [comment]
This works, as long as the application process isn't killed by the system to reclaim memory.
What it means is that the application can still be restored but `isInBackground` will be reset and the sync won't trigger.

We could add another flag there to trigger `backgroundService.EnterBackground();` when the first activity is resumed after the application is created.

You can simulate the process being killed by going into developer options > Apps > Background process limit.

**Comment by juanlaube** [comment]
@daividssilverio `isInBackground` starts in `true` so whenever the app is created `EnterForeground` will be called. This should work regardless of the app being just launched or returning from a tombstoned state.

**Comment by daividssilverio** [comment]
Then something else is breaking it... I'm always getting `Sync Failed` when returning to a killed process.

**Comment by juanlaube** [comment]
API token maybe? The fact that it tries to sync, even if it fails, means this work tho 👀

**Comment by daividssilverio** [comment]
I see why it won't trigger the sync, the `isInBackground` starts with true, but the AppResumedFromBackground will emit a value less than `minimumTimeInBackgroundForFullSync`, if you check `TogglDataSource.cs`

**Comment by juanlaube** [comment]
Re the sync failures, I think it's better to follow up this in a new issue. I created #4047 for that.

**Comment by juanlaube** [comment]
@simonrozsival sync fails sometimes only when the app is tombstoned by the OS (you can limit the active processes to test this), in a normal situation this is not a problem.

**Comment by juanlaube** [comment]
The follow up issue has been closed. For some reason there's a problem with the self signed certificate but that only happens in debug builds, so all is good for now.

**Pull request merged**